internal/core/src/common/GeoSpatial.h

```diff
@@ -0,0 +1,156 @@
+            this->geometry_ = other.geometry_->clone();
+            this->to_wkb_internal();
+        }
+    }
+
+    GeoSpatial(GeoSpatial&& other) noexcept
+        : wkb_data_(std::move(other.wkb_data_)),
+          size_(std::move(other.size_)),
+          geometry_(std::move(other.geometry_)) {
+    }
+
+    GeoSpatial&
+    operator=(const GeoSpatial& other) {
+        if (this != &other && other.IsValid()) {
+            this->geometry_ = other.geometry_->clone();
+            this->to_wkb_internal();
+        }
+        return *this;
+    }
+
+    GeoSpatial&
+    operator=(GeoSpatial&& other) noexcept {
+        if (this != &other) {
+            wkb_data_ = std::move(other.wkb_data_);
+            size_ = std::move(other.size_);
+            geometry_ = std::move(other.geometry_);
+        }
+        return *this;
+    }
+
+    operator std::string() const {
+        //tmp string created by copy ctr
+        return std::string(reinterpret_cast<const char*>(wkb_data_));
+    }
+
+    operator std::string_view() const {
+        return std::string_view(reinterpret_cast<const char*>(wkb_data_),
+                                size_);
+    }
+
+    ~GeoSpatial() {
+        if (geometry_) {
+            OGRGeometryFactory::destroyGeometry(geometry_);
+        }
+        if (wkb_data_) {  // the data
+            delete[] wkb_data_;
+        }
+    }
+
+    bool
+    IsValid() const {
+        return geometry_ != nullptr;
+    }
+
+    OGRGeometry*
+    GetGeometry() const {
+        return geometry_;
+    }
+
+    const unsigned char*
+    data() const {
+        return wkb_data_;
+    }
+
+    size_t
+    size() const {
+        return size_;
+    }
+
+    //spatial relation
+    bool
+    equals(const GeoSpatial& other) const {
+        return geometry_->Equals(other.geometry_);
+    }
+
+    bool
+    touches(const GeoSpatial& other) const {
+        return geometry_->Touches(other.geometry_);
+    }
+
+    bool
+    overlaps(const GeoSpatial& other) const {
+        return geometry_->Overlaps(other.geometry_);
+    }
+
+    bool
+    crosses(const GeoSpatial& other) const {
+        return geometry_->Crosses(other.geometry_);
+    }
+
+    bool
+    contains(const GeoSpatial& other) const {
+        return geometry_->Contains(other.geometry_);
+    }
+
+    bool
+    intersects(const GeoSpatial& other) const {
+        return geometry_->Intersects(other.geometry_);
+    }
+
+    bool
+    within(const GeoSpatial& other) const {
+        return geometry_->Within(other.geometry_);
+    }
+
+ private:
+    inline void
+    to_wkb_internal() {
+        if (geometry_ && size_ == 0) {
+            size_ = geometry_->WkbSize();
+            wkb_data_ = new unsigned char[size_];
+            // little-endian order to save wkb
+            geometry_->exportToWkb(wkbNDR, wkb_data_);
+        }
+    }
+
+    //read write ptr, use to save a OGRGeometry object when need to create new storage file
+    unsigned char* wkb_data_{nullptr};
```

Comment:
can `std::unique_ptr<unsigned char[]>` be used here instead of a raw pointer?

Reply:
Thanks for your suggestions.